src/module.c

```diff
@@ -8365,14 +8374,22 @@ void moduleBlockedClientTimedOut(client *c) {
     if (bc->unblocked) return;
 
     RedisModuleCtx ctx;
-    moduleCreateContext(&ctx, bc->module, REDISMODULE_CTX_BLOCKED_TIMEOUT);
+    int flags = REDISMODULE_CTX_BLOCKED_TIMEOUT;
+    if (from_module) flags |= REDISMODULE_CTX_THREAD_SAFE;
```

Comment:
This is because both of them are fixed to ensure that `moduleBlockedClientTimedOut()` is thread-safe.

The reason is:
```
When error reply is called in timeout_callback(), ctx is not thread-safe, eventually lead to race conditions in afterErrorReply.
```

If the ctx is not to be `REDISMODULE_CTX_THREAD_SAFE`, `afterErrorReply()` will be triggered if  replying an error.
If `REDISMODULE_CTX_THREAD_SAFE` is used, all replies will be deferred.
(https://github.com/redis/redis/blob/27a8e3b04e645e0c614c597668355e9bc945c408/src/networking.c#L501-L507)